### PR TITLE
Shim controller

### PR DESCRIPTION
**Pull request by melkobrodov**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #2064 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation of turtlebot3 |

---

## Description of contribution in a few bullet points

* I added shim controller plugin

## Description of documentation updates required from your changes

* I added parameters, need to document them in Configuration Guide section
* Also I added tutorial in General Tutorials

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_bringup/params/nav2_sc_params.yaml [hunk 1]
```diff
@@ -0,0 +1,341 @@
+amcl:
```
Comment (SteveMacenski) [inline nav2_bringup/params/nav2_sc_params.yaml:1]: Remove this param file

File: nav2_shim_controller/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+find_package(geometry_msgs REQUIRED)
+find_package(nav2_costmap_2d REQUIRED)
+find_package(pluginlib REQUIRED)
+find_package(nav_msgs REQUIRED)
+find_package(nav2_util REQUIRED)
+find_package(nav2_core REQUIRED)
+find_package(tf2 REQUIRED)
+
+nav2_package()
+set(CMAKE_CXX_STANDARD 17)
```
Comment (SteveMacenski) [inline nav2_shim_controller/CMakeLists.txt:21]: This shouldn't be necessary, what C++17 feature are you using?

File: nav2_shim_controller/include/nav2_shim_controller/shim_controller.hpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2020 Shrijit Singh
```
Comment (SteveMacenski) [inline nav2_shim_controller/include/nav2_shim_controller/shim_controller.hpp:1]: This package is missing a readme file explaining what it is, how to use it, why you want to use it, parameters in it, etc.

File: nav2_shim_controller/include/nav2_shim_controller/shim_controller.hpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    const geometry_msgs::msg::PoseStamped & pose,
+    const geometry_msgs::msg::Twist & velocity);
+
+  std::shared_ptr<tf2_ros::Buffer> tf_;
+  std::string plugin_name_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros_;
+  nav2_costmap_2d::Costmap2D * costmap_;
+  rclcpp::Logger logger_ {rclcpp::get_logger("ShimController")};
+
+  double desired_linear_vel_;
```
Comment (SteveMacenski) [inline nav2_shim_controller/include/nav2_shim_controller/shim_controller.hpp:189]: Many of these parameters should not exist, there's no need for this param, or look ahead dist, or most of them below. This controller only needs a param for max angular velocity / acceleration for rotating to heading / path.

File: nav2_shim_controller/src/shim_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+void ShimController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type"
+    " shim_controller::ShimController",
+    plugin_name_.c_str());
+
+  if (default_plugin) {default_plugin->cleanup();}
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:125]: Expand onto multiple lines (and for all cases of this below) to match styling of the rest of the program.

Comment (SteveMacenski) [reply nav2_shim_controller/src/shim_controller.cpp:125]: Also need to `reset()` the pointer, I believe in cleanup.

File: nav2_shim_controller/src/shim_controller.cpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+geometry_msgs::msg::TwistStamped ShimController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed,
+  nav2_core::GoalChecker * goal_checker)
+{
+  goal_checker_ = goal_checker;
+  return calcCmdVel(pose, speed);
+}
+
+geometry_msgs::msg::TwistStamped ShimController::calcCmdVel(
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:158]: Put this function into `computeVelocityCommands`, there's no real reason to separate this out.

Comment (SteveMacenski) [reply nav2_shim_controller/src/shim_controller.cpp:158]: Then goal_checker doesn't need to be a class member anymore

File: nav2_shim_controller/src/shim_controller.cpp [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+{
+  global_plan_ = path;
+  if (default_plugin) {default_plugin->setPlan(global_plan_);}
+}
+
+void ShimController::setSpeedLimit(
+  const double & speed_limit,
+  const bool & percentage)
+{
+  if (speed_limit == nav2_costmap_2d::NO_SPEED_LIMIT) {
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:295]: This should all just be a pass through to the sub-plugin, there's no need for any of this logic here.

File: nav2_shim_controller/src/shim_controller.cpp [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+  RCLCPP_DEBUG(
+    logger_, "speed = %f, speed_cur = %f, angle = %f, calc angle = %f",
+    angular_vel,
+    curr_speed.twist.angular.z,
+    angle_to_path,
+    deceleration_angle);
+}
+
+geometry_msgs::msg::PoseStamped ShimController::getLookAheadPoint(
+  const double & lookahead_dist,
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:268]: This function is unneeded, remove

File: nav2_shim_controller/src/shim_controller.cpp [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    } else {
+      // Speed limit is expressed in absolute value
+      desired_linear_vel_ = speed_limit;
+    }
+  }
+  if (default_plugin) {default_plugin->setSpeedLimit(speed_limit, percentage);}
+}
+
+nav_msgs::msg::Path ShimController::transformGlobalPlan(
+  const geometry_msgs::msg::PoseStamped & pose)
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:311]: This function is unneeded, remove

File: nav2_shim_controller/src/shim_controller.cpp [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+  if (transformed_plan.poses.empty()) {
+    throw nav2_core::PlannerException("Resulting plan has 0 poses in it.");
+  }
+
+  return transformed_plan;
+}
+
+bool ShimController::transformPose(
+  const std::string frame,
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:373]: This function is unneeded, remove

File: nav2_shim_controller/src/shim_controller.cpp [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+  geometry_msgs::msg::Twist vel_tolerance;
+  if (goal_checker_ && goal_checker_->getTolerances(pose_tolerance, vel_tolerance)) {
+    goal_dist_tol_ = pose_tolerance.position.x;
+    goal_yaw_tol_ = tf2::getYaw(pose_tolerance.orientation);
+  } else {
+    RCLCPP_WARN(logger_, "Unable to retrieve goal checker's tolerances!");
+  }
+
+  // Transform path to robot base frame
+  auto transformed_plan = transformGlobalPlan(pose);
```
Comment (SteveMacenski) [inline nav2_shim_controller/src/shim_controller.cpp:173]: Remove all RPP logic, I don't see why that should be in the file that simply rotates a robot to its current path / goal headings

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Code looks clean for the most part, just need to pare it down to the necessary minimum components and simplify it. Our goal is just a plugin that rotates a robot to the rough path heading once a new path comes in, and when at the goal pose. Otherwise, give it to the controller plugin to track the path. Overall, this should be a relatively simple passthrough file with some block of rotate to heading logic.

File: nav2_shim_controller/src/shim_controller.cpp [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+    desired_linear_vel_ = max_linear_vel_;
+  } else {
+    if (percentage) {
+      // Speed limit is expressed in % from maximum speed of robot
+      desired_linear_vel_ = max_linear_vel_ * speed_limit / 100.0;
+    } else {
+      // Speed limit is expressed in absolute value
+      desired_linear_vel_ = speed_limit;
+    }
+  }
```
Comment (AlexeyMerzlyakov) [inline nav2_shim_controller/src/shim_controller.cpp:306]: Where is this logic used? It looks, like `desired_linear_vel_` is unused at all: we could just call `default_plugin->setSpeedLimit()`, I think

**Review by AlexeyMerzlyakov — COMMENTED** [review]
I would like agree with Steve's main comment: it looks like we could do without `transformGlobalPlan()/getLookAheadPoint()` PRR's functions as knowing current robot pose and global path should be enough for decisions making to rotate and won't break any current logic, as currently it relies on predicted `carrot_pose` on one step ahead which could be calculated as tangent to global path.

File: nav2_shim_controller/src/shim_controller.cpp (hunk 11 above)
Comment (SteveMacenski) [reply nav2_shim_controller/src/shim_controller.cpp:173]: The only time you need to do any of this logic is when a new path comes in. So in the set plan function, you could set a flag that it has changed (or compare the paths to see that there is a change) to then trigger the "check if I even need to rotate to rough heading" logic. If not, just pass through and reset the flag. If so, rotate to heading. Once rotated within the tolerance, reset the flag as pass through to the internal controller plugin.

**Comment by SteveMacenski** [comment]
Hi,

Unfortunately, this is needed a bit more immediately for a project so I'm planning on tackling this this week in order to get it done and merged in before year's end. I'll leave this PR open in case you get a chance to fix it before I'm done, but I'm targeting Friday for a PR to resolve this feature request.

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation2/pull/2718

**Pull request closed without merging**